simapp/app.go

```diff
@@ -450,7 +451,9 @@ func (app *SimApp) EndBlocker(ctx sdk.Context, req abci.RequestEndBlock) abci.Re
 // InitChainer application update at chain initialization
 func (app *SimApp) InitChainer(ctx sdk.Context, req abci.RequestInitChain) abci.ResponseInitChain {
 	var genesisState GenesisState
-	app.cdc.MustUnmarshalJSON(req.AppStateBytes, &genesisState)
+	if err := tmjson.Unmarshal(req.AppStateBytes, &genesisState); err != nil {
```

Comment:
it's a pity that the library does not provide a `MustUnmarshal()` method :(

Reply:
I see it as an advantage. Panics should be avoided.

Reply:
what is the reason for using tmjson here?

Reply:
I thought GenesisDoc was marshalled using tmjson on the TM side?

Reply:
ah wait, i might have read too fast, this is our own GenesisState. So it should be decoded with appCodec.

Reply:
Witin Tendermint this is `json.RawMessage`, I thought the app's json decoder would handle it, I may be wrong if there are tests that check this? (ci passed)

Reply:
I just double checked: GenesisState is a `map[string]json.RawMessage`. So a json.Unmarshal is enough. The actual decoding using appCodec is done in InitGenesis (2 lines below).

I'll create a PR for this

Reply:
> I see it as an advantage. Panics should be avoided.

Sure. When the error can be handled correctly and program's state recovered fully. Else you want to panic and produce as much logging information as possible. So the real question is: how do you recover from encountering a bad encoded state? If we have an answer to that, then I agree and we can avoid using `panic()` here.

Reply:
Proper error handling:

* define critical error type
* pass it up with a stack
* handle it in the main routine allowing to shutdown things (maybe some shutdown's are not relevant here, but this is the right approach anyway).

Reply:
> pass it up with a stack

I +1 this idea, and would even propose more aggressively to only allow Tendermint to panic. It's probably a larger discussion, but if the TM spec would expose some well-defined errors that cause TM to panic, then the SDK (or any state machine) should only return these errors, and contain 0 panic itself.

Reply:
Two weeks ago I've created an issue to clean the panics wherever possible: https://github.com/cosmos/cosmos-sdk/issues/7409

Reply:
> It's probably a larger discussion

I agree very much indeed to that. But until we got stuff sorted in Tendermint, there is not much we can do other than `panic()`ing in the SDK.

Reply:
We generally have been using stdlib encoding/json